ember-server/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpers.scala

```diff
@@ -0,0 +1,217 @@
+            // TODO: improve
+            val bytes = new Array[Byte](buffer.remaining())
+            buffer.get(bytes)
+            Chunk.bytes(bytes)
+          })
+          .flatMap(Stream.chunk(_))
+      }
+      .through(socket.writes(durationToFinite(idleTimeout)))
+
+    val reader = (Stream.chunk(Chunk.bytes(buffer)) ++ readStream(read))
+      .through(decodeFrames(frameTranscoder))
+      .through(ctx.webSocket.receive)
+
+    reader
+      .concurrently(writer)
+      .drain
+      .compile
+      .drain
+      .attempt
+      .flatMap {
+        case Left(err) =>
+          err.printStackTrace()
+          F.unit
+        case Right(()) =>
+          println("Connection ended")
+          F.unit
+      }
+  }
+
+  private def decodeFrames[F[_]](frameTranscoder: FrameTranscoder)(implicit
+      F: Concurrent[F]): Pipe[F, Byte, WebSocketFrame] = stream => {
+    def go(rest: Stream[F, Byte], acc: Array[Byte]): Pull[F, WebSocketFrame, Unit] =
+      rest.pull.uncons.flatMap {
+        case Some((chunk, next)) =>
+          val buffer = acc ++ chunk.toArray[Byte]
+          val byteBuffer = ByteBuffer.wrap(buffer)
+          Pull
+            .attemptEval(F.delay(frameTranscoder.bufferToFrame(byteBuffer)))
+            .flatMap {
+              case Right(value) =>
+                // TODO: improve this buffering
+                if (value != null) {
+                  val remaining = new Array[Byte](byteBuffer.remaining())
+                  byteBuffer.get(remaining)
+                  Pull.output1(value) >> go(next, remaining)
+                } else {
+                  go(next, buffer)
+                }
+              case Left(err) =>
+                // TODO: figure out what to do here
+                println(err)
+                Pull.done
+            }
+        case None =>
+          // TODO: figure out what to do here
+          println("done")
+          Pull.done
+      }
+
+    go(stream, Array.emptyByteArray).void.stream
+  }
+
+  private def clientHandshake[F[_]](req: Request[F]): Either[ClientHandshakeError, String] = {
+    val connection = req.headers.get(Connection) match {
+      case Some(header) if header.values.contains_(upgradeCi) => Right(())
+      case _ => Left(UpgradeRequired)
+    }
+
+    val upgrade = req.headers.get(Upgrade) match {
+      case Some(header) if header.values.contains_(webSocketCi) => Right(())
+      case _ => Left(UpgradeRequired)
+    }
+
+    val version = req.headers.get(`Sec-WebSocket-Version`) match {
+      case Some(header) if header.version == supportedWebSocketVersion => Right(())
+      case Some(header) => Left(UnsupportedVersion(supportedWebSocketVersion, header.version))
+      case None => Left(VersionNotFound)
+    }
+
+    val key = req.headers.get(`Sec-WebSocket-Key`) match {
+      case Some(header) => Right(header.value)
+      case None => Left(KeyNotFound)
+    }
+
+    (connection, upgrade, version, key).mapN { case (_, _, _, key) => key }
```

Comment:
Should this be a `Validated` so we can list all our complaints?

Reply:
I don't think the errors are mutually independent. For example, does it make sense to check the version or key headers if the connection and upgrade headers aren't right? The response code might also be different in the case of the former two checks (426 upgrade required)